Add K/Wasm support

eymar opened the pull request:
This PR:
- updates kotlin to 1.9.21
- removes watchosX86() target - it's not available since kotlin 1.9.20

eymar committed:
Include skiko.wasm in tests (ui, foundation, etc) for k/wasm target

eymar committed:
Add missing compose test API for k/wasm target (#553)

eymar committed:
Workaround https://youtrack.jetbrains.com/issue/KT-58293

eymar committed:
Implement PlatformLocale and StringDelegate for k/wasm (#561)

* Workaround `isOnUiThread` (ui-test) for k/wasm browser target

* Implement PlatformLocale  and StringDelegate for k/wasm

* Get rid of expect/actuals for jsWasmMain and use js(...) for a common implementation

* use the list of preferred languages - `window.navigator.languages`

* delete unused fun `currentLanguageTag`

eymar committed:
add compose:mpp:demo run configuration for k/wasm target

eymar committed:
Update kotlin to 1.9.0 and update other libs

eymar committed:
Fix publication

eymar committed:
Sync 1. Kotlin 1.9.0

eymar committed:
enable k/wasm (not working yet)

eymar committed:
enable k/wasm 2 (not working yet)

eymar committed:
Refactoring of k/wasm sources

eymar committed:
k/wasm demo runs

eymar committed:
publication fixes

eymar committed:
Fixes

eymar committed:
Update kotlin to 1.9.20-RC

eymar committed:
Update oel.androidx.version to 1.5.4

eymar committed:
Update kotlin to 1.9.20 and new libs

eymar committed:
Remove deleted AllowDifferentMembersInActual annotation usages

eymar committed:
Rebase fixes

eymar committed:
update to kotlin 1.9.21

eymar committed:
cleanup

eymar committed:
cleanup

eymar committed:
cleanup

eymar committed:
keep generateDecoys for k/js, but not for k/wasm

eymar committed:
cleanup

eymar committed:
make tests run for k/wasm target

eymar commented on buildSrc/plugins/src/main/kotlin/androidx/build/AndroidXPlugin.kt:
```diff
@@ -35,6 +35,19 @@ class AndroidXPlugin : Plugin<Project> {
         )
+
+        project.configurations.all {
+            // TODO: remove these HACKS when possible
```
It's not supposed to be merged into jb-main branch.
Should be removed when all required libs are properly built and published

eymar commented on buildSrc/private/src/main/kotlin/androidx/build/AndroidXComposeImplPlugin.kt:
```diff
@@ -138,10 +138,20 @@ class AndroidXComposeImplPlugin : Plugin<Project> {
             }
 
             project.tasks.withType(KotlinJsCompile::class.java).configureEach { compile ->
+                val isWasm = compile.kotlinOptions.freeCompilerArgs.contains("-Xwasm")
```
Currently KotlinJsCompile task type is reused for k/wasm compilations.

For the nearest release we'll keep decoys enabled for k/js, but not for k/wasm. In the future it will be disabled for both.

eymar commented on buildSrc/private/src/main/kotlin/androidx/build/AndroidXComposeImplPlugin.kt:
```diff
@@ -366,17 +376,7 @@ private fun configureComposeCompilerPlugin(
-                project.rootProject.resolveProject(":compose:compiler:compiler")
-            }
-        )
+        project.dependencies.add(COMPILER_PLUGIN_CONFIGURATION, "org.jetbrains.compose.compiler:compiler:1.5.4")
```
To build the compose-core libs with our published compose compiler plugin, rather than from the plugin built from the sources.
This way we have a single source of truth.

igordmn replied:
let's move this version to `gradle.properties`

eymar replied:
done

eymar commented on compose/compiler/compiler-hosted/src/main/java/androidx/compose/compiler/plugins/kotlin/ComposeIrGenerationExtension.kt:
```diff
@@ -125,6 +126,7 @@ class ComposeIrGenerationExtension(
 
         val mangler = when {
             pluginContext.platform.isJs() -> JsManglerIr
+            pluginContext.platform.isWasm() -> JsManglerIr
```
In the future, the changes in compose/compiler can be ignored. When merging from upstream we can take their sources. 
We have a separate branch with our patches for building and publishing the compose compiler plugin.

eymar commented on testutils/testutils-kmp/src/commonMain/kotlin/androidx/kruth/Subject.kt:
```diff
@@ -193,6 +193,6 @@ private fun Any?.integralValue(): Long = when (this) {
 private fun Any?.toStringForAssert(): String = when {
     this == null -> toString()
-    isIntegralBoxedPrimitive() -> "${this::class.qualifiedName}<$this>"
+    isIntegralBoxedPrimitive() -> "${this::class.simpleName}<$this>"
```
`qualifiedName` is a reflection API not available for web targets.

eymar commented on compose/mpp/demo-uikit/build.gradle.kts:
```diff
@@ -112,20 +112,3 @@ apple {
 }
-
-// TODO: Workaround, see https://youtrack.jetbrains.com/issue/KT-55751
-val myAttribute = Attribute.of("myOwnAttribute", String::class.java)
```
Dima has checked that this workaround is not needed anymore with kotlin 1.9.2x

eymar committed:
Use 0.0.0-SNAPSHOT skiko to get rid of demo extra configuration

eymar commented on compose/ui/ui-test-junit4/src/skikoMain/kotlin/androidx/compose/ui/test/junit4/ComposeRootRegistry.skiko.kt:
```diff
@@ -29,18 +29,22 @@ internal class ComposeRootRegistry {
      */
     val isSetUp: Boolean
-        get() = SkiaRootForTest.onRootCreatedCallback == ::onRootCreated
+        get() = SkiaRootForTest.onRootCreatedCallback == _onRootCreated
```
cc @MatkovIvan JFYI: unfortunately, not all targets reuse the same reference here. 
Apparently, k/wasm creates a new instance every time and the check is always false.

MatkovIvan replied:
Let's just store `isSetUp` as a boolean flag then

eymar replied:
ACK

igordmn commented on .gitignore:
```diff
@@ -39,3 +39,5 @@ __pycache__
 
 # Don't ignore src files under build package
 !**/src/**/build
+/jbdeps/android-sdk/linux/*
```
Why do we need this? We already have `jbdeps/.gitignore`

eymar replied:
Hm. Right, for some reason it didn't ignore the files for me.

eymar replied:
removed the change

igordmn commented on buildSrc/build.gradle:
```diff
@@ -32,7 +32,7 @@ allprojects {
             jvmTarget = "17"
             freeCompilerArgs += [
-                    "-Werror",
+                    // "-Werror", // commented in JB-fork
```
Can we not comment this, and fix the warnings?

eymar replied:
done

igordmn commented on compose/runtime/runtime/src/nativeMain/kotlin/androidx/compose/runtime/ActualNative.native.kt:
```diff
@@ -59,19 +60,21 @@ actual class AtomicReference<V> actual constructor(value: V) {
     actual fun add(amount: Int): Int = delegate.addAndGet(amount)
 }
 
+@OptIn(ExperimentalNativeApi::class)
```
We should keep in mind that we can't release Compose for iOS if at that time it is still be experimental. It doesn't add binary compatibility guarantees, which we need for stable Compose targets.

igordmn commented on compose/runtime/runtime/src/wasmJsTest/kotlin/kotlinx/test/Actuals.wasm.kt:
```diff
@@ -0,0 +1,26 @@
+
+@OptIn(DelicateCoroutinesApi::class)
+actual suspend fun testWithTimeout(timeoutMs: Long, block: suspend CoroutineScope.() -> Unit) {
+    GlobalScope.launch {
```
```suggestion
    coroutineScope {
         block()
    }
```
Will this work? An async call is fragile.

eymar replied:
This function is not implemented properly at all (I added this dumb implementation a few month ago and forgot about it). It ignores `timeoutMs`. I'll update it separately. 
It's used in our tests only.

eymar replied:
COMPOSE-662

igordmn commented on compose/ui/ui-test-junit4/src/wasmJsMain/kotlin/androidx/compose/ui/test/junit4/Synchronization.wasmMain.kt:
```diff
@@ -0,0 +1,42 @@
+ */
+
+// TODO: in a browser we have only 1 thread and it's actually a UI thread.
+// But returning `true` here breaks `setContent` - `waitForIdle` is not called, but seems to be necessary.
```
Let's create an issue to investigate this. It shouldn't work this way.

eymar replied:
Added to the comment - COMPOSE-661

eymar committed:
fixes for commonMain metadata compilation

eymar committed:
Use an official kotlinSerialization version with k/wasm support

eymar committed:
Disable decoys for k/js

eymar committed:
Update runtime.api file

reason: I guess new compiler plugin generated added stable field to CompositionScopedCoroutineScopeCanceller. Also new kotlin added getEntries method to Recomposer$State

eymar committed:
Ignore SnapshotStateMapTests.validateValuesRemove for js

eymar committed:
PR review fixes

eymar committed:
Update more .api files: new `getEntries` method for enums

eymar committed:
Fix runtime tests

eymar committed:
Make PlatformDateFormat.jsWasm.kt common for k/s and k/wasm in material3

eymar commented on buildSrc/build.gradle:
```diff
@@ -39,8 +39,8 @@ allprojects {
-            languageVersion = "1.5"
-            apiVersion = "1.5"
+            languageVersion = "1.8"
+            apiVersion = "1.8"
```
kotlin 1.9.21 shows a warning for 1.5

eymar committed:
remove wasm/experimental maven repositories

eymar committed:
Fixes after rebase

eymar committed:
add YT issue

eymar committed:
add YT issue
